Qbeast does not support data type java.time.Instant (Hive)

## What went wrong?
Indexing timestamp fields with Qbeast can lead to an error when using Hive (possibly other data warehouses too). 
`scala.MatchError: 2022-09-30T12:53:19Z (of class java.time.Instant)`
This bug is not reproducible with spark-shell. So it is possible that the problem comes from the way that Hive treats timestamps internally, which is different than spark SQL.

## How to reproduce?

Read or create a timestamp field and index the field with qbeast format. This bug is reproducible when using Hive SQL (beeline)

### 1. Code that triggered the bug, or steps to reproduce:

1. Start thrift server:
```bash
./spark-3.2.2-bin-hadoop3.2/sbin/start-thriftserver.sh --conf spark.sql.extensions=io.qbeast.spark.internal.QbeastSparkSessionExtension --conf spark.sql.catalog.spark_catalog=io.qbeast.spark.internal.sources.catalog.QbeastCatalog
```

2. Run beeline:
```bash
./spark-3.2.2-bin-hadoop3.2/bin/beeline -u jdbc:hive2://localhost:10000
```
```sql
CREATE DATABASE test;
CREATE TABLE test.source VALUES 1664542399, 1664542399;
CREATE TABLE test.time using qbeast options ("columnsToIndex"="col1", "cubeSize"="5000");
INSERT INTO test.time SELECT cast(col1 as timestamp) from test.source;
```

### 2. Branch and commit id:

Main branch - commit: 014aa0e882a605c332dc823bb027ffad835627bd

### 3. Spark version: 3.2.2

### 4. Hadoop version: 3.2

### 5. How are you running Spark?

Using spark through thrift server and Hive SQL on local machine.

### 6. Stack trace:

```
0: jdbc:hive2://localhost:10000> INSERT INTO test.time SELECT cast(col1 as timestamp) from test.source;
Error: org.apache.hive.service.cli.HiveSQLException: Error running query: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 18.0 failed 1 times, most recent failure: Lost task 0.0 in stage 18.0 (TID 19) (192.168.1.46 executor driver): scala.MatchError: 2022-09-30T12:53:19Z (of class java.time.Instant)
	at io.qbeast.core.transform.IdentityToZeroTransformation.transform(Transformation.scala:54)
	at io.qbeast.spark.index.RowUtils$.$anonfun$rowValuesToPoint$1(RowUtils.scala:30)
	at io.qbeast.spark.index.RowUtils$.$anonfun$rowValuesToPoint$1$adapted(RowUtils.scala:28)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at io.qbeast.spark.index.RowUtils$.rowValuesToPoint(RowUtils.scala:28)
	at io.qbeast.spark.index.DoublePassOTreeDataAnalyzer$.$anonfun$estimatePartitionCubeWeights$5(OTreeDataAnalyzer.scala:153)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at io.qbeast.spark.index.DoublePassOTreeDataAnalyzer$.$anonfun$estimatePartitionCubeWeights$4(OTreeDataAnalyzer.scala:152)
	at org.apache.spark.sql.execution.MapPartitionsExec.$anonfun$doExecute$3(objects.scala:201)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:898)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:898)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1491)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)

Driver stacktrace:
	at org.apache.spark.sql.hive.thriftserver.HiveThriftServerErrors$.runningQueryError(HiveThriftServerErrors.scala:43)
	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation.org$apache$spark$sql$hive$thriftserver$SparkExecuteStatementOperation$$execute(SparkExecuteStatementOperation.scala:325)
	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2$$anon$3.$anonfun$run$2(SparkExecuteStatementOperation.scala:230)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.sql.hive.thriftserver.SparkOperation.withLocalProperties(SparkOperation.scala:79)
	at org.apache.spark.sql.hive.thriftserver.SparkOperation.withLocalProperties$(SparkOperation.scala:63)
	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation.withLocalProperties(SparkExecuteStatementOperation.scala:43)
	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2$$anon$3.run(SparkExecuteStatementOperation.scala:230)
	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2$$anon$3.run(SparkExecuteStatementOperation.scala:225)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1878)
	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2.run(SparkExecuteStatementOperation.scala:239)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 18.0 failed 1 times, most recent failure: Lost task 0.0 in stage 18.0 (TID 19) (192.168.1.46 executor driver): scala.MatchError: 2022-09-30T12:53:19Z (of class java.time.Instant)
	at io.qbeast.core.transform.IdentityToZeroTransformation.transform(Transformation.scala:54)
	at io.qbeast.spark.index.RowUtils$.$anonfun$rowValuesToPoint$1(RowUtils.scala:30)
	at io.qbeast.spark.index.RowUtils$.$anonfun$rowValuesToPoint$1$adapted(RowUtils.scala:28)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at io.qbeast.spark.index.RowUtils$.rowValuesToPoint(RowUtils.scala:28)
	at io.qbeast.spark.index.DoublePassOTreeDataAnalyzer$.$anonfun$estimatePartitionCubeWeights$5(OTreeDataAnalyzer.scala:153)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at io.qbeast.spark.index.DoublePassOTreeDataAnalyzer$.$anonfun$estimatePartitionCubeWeights$4(OTreeDataAnalyzer.scala:152)
	at org.apache.spark.sql.execution.MapPartitionsExec.$anonfun$doExecute$3(objects.scala:201)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:898)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:898)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1491)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2454)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2403)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2402)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2402)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1160)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1160)
	at scala.Option.foreach(Option.scala:407)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1160)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2642)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2584)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2573)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
Caused by: scala.MatchError: 2022-09-30T12:53:19Z (of class java.time.Instant)
	at io.qbeast.core.transform.IdentityToZeroTransformation.transform(Transformation.scala:54)
	at io.qbeast.spark.index.RowUtils$.$anonfun$rowValuesToPoint$1(RowUtils.scala:30)
	at io.qbeast.spark.index.RowUtils$.$anonfun$rowValuesToPoint$1$adapted(RowUtils.scala:28)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at io.qbeast.spark.index.RowUtils$.rowValuesToPoint(RowUtils.scala:28)
	at io.qbeast.spark.index.DoublePassOTreeDataAnalyzer$.$anonfun$estimatePartitionCubeWeights$5(OTreeDataAnalyzer.scala:153)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at io.qbeast.spark.index.DoublePassOTreeDataAnalyzer$.$anonfun$estimatePartitionCubeWeights$4(OTreeDataAnalyzer.scala:152)
	at org.apache.spark.sql.execution.MapPartitionsExec.$anonfun$doExecute$3(objects.scala:201)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:898)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:898)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1491)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829) (state=,code=0)
```

--- a/core/src/main/scala/io/qbeast/core/transform/LinearTransformation.scala
+++ b/core/src/main/scala/io/qbeast/core/transform/LinearTransformation.scala
@@ -11,18 +11,19 @@ import com.fasterxml.jackson.databind.node.{DoubleNode, IntNode, NumericNode, Te
 import com.fasterxml.jackson.databind.ser.std.StdSerializer
 import com.fasterxml.jackson.databind.{DeserializationContext, SerializerProvider}
 import io.qbeast.core.model.{
+  DateDataType,
   DecimalDataType,
   DoubleDataType,
   FloatDataType,
   IntegerDataType,
   LongDataType,
   OrderedDataType,
-  TimestampDataType,
-  DateDataType
+  TimestampDataType
 }
 
 import java.math.BigDecimal
-import java.sql.{Timestamp, Date}
+import java.sql.{Date, Timestamp}
+import java.time.Instant
 import scala.util.Random
 import scala.util.hashing.MurmurHash3
 
@@ -62,6 +63,7 @@ case class LinearTransformation(
       case v: Float => (v - mn) * scale
       case v: Timestamp => (v.getTime - mn) * scale
       case v: Date => (v.getTime - mn) * scale
+      case v: Instant => (v.toEpochMilli - mn) * scale
     }
   }
 

--- a/core/src/main/scala/io/qbeast/core/transform/LinearTransformer.scala
+++ b/core/src/main/scala/io/qbeast/core/transform/LinearTransformer.scala
@@ -29,7 +29,7 @@ case class LinearTransformer(columnName: String, dataType: QDataType) extends Tr
       case d: java.math.BigDecimal => d.doubleValue()
       case d: Timestamp => d.getTime
       case d: Date => d.getTime
-      case d: Instant => Timestamp.from(d).getTime
+      case d: Instant => d.toEpochMilli
       case other => other
     }
   }

--- a/core/src/main/scala/io/qbeast/core/transform/LinearTransformer.scala
+++ b/core/src/main/scala/io/qbeast/core/transform/LinearTransformer.scala
@@ -6,6 +6,7 @@ package io.qbeast.core.transform
 import io.qbeast.core.model.{OrderedDataType, QDataType}
 
 import java.sql.{Date, Timestamp}
+import java.time.Instant
 
 object LinearTransformer extends TransformerType {
   override def transformerSimpleName: String = "linear"
@@ -26,8 +27,9 @@ case class LinearTransformer(columnName: String, dataType: QDataType) extends Tr
       // Very special case in which we load the transformation information from JSON options
       case d: java.lang.Long if dataType.name == "IntegerDataType" => d.intValue()
       case d: java.math.BigDecimal => d.doubleValue()
-      case d: Timestamp => d.getTime()
-      case d: Date => d.getTime()
+      case d: Timestamp => d.getTime
+      case d: Date => d.getTime
+      case d: Instant => Timestamp.from(d).getTime
       case other => other
     }
   }

--- a/core/src/main/scala/io/qbeast/core/transform/Transformation.scala
+++ b/core/src/main/scala/io/qbeast/core/transform/Transformation.scala
@@ -4,7 +4,9 @@
 package io.qbeast.core.transform
 
 import com.fasterxml.jackson.annotation.JsonTypeInfo
-import java.sql.{Timestamp, Date}
+
+import java.sql.{Date, Timestamp}
+import java.time.Instant
 
 /**
  * Double value transformation.
@@ -56,6 +58,7 @@ case class IdentityToZeroTransformation(identityValue: Any) extends Transformati
     case v: Number if v == identityValue => 0.0
     case v: Timestamp if v == identityValue => 0.0
     case v: Date if v == identityValue => 0.0
+    case v: Instant if v == identityValue => 0.0
 
   }
 

